README.md
python-minecraft
================

This repository contains example Python scripts for use with Minecraft at [CoderDojo Twin Cities](http://www.coderdojotc.org/).

The actual MCPi API v0.1_1 has been unceremoniously included in this repository in the interest of getting students up and running quickly. It was obtained from [pi.minecraft.net](http://pi.minecraft.net/).

Go [here](/SETUP_README.md) for instructions on how to get the rest of your environment up and running.

Coordinate System
------------------
Most coordinates are in the form of a three integer vector (x,y,z) which address a specific tile in the game world. (0,0,0) is the spawn point sea level. (x,z) is the ground plane and Y is towards the sky. In other words, x is left and right, z is forward and backward, and y is up and down.

![Minecraft's odd x-y-z coordinate system](/ignoreme/coordinates.png)

Minecraft Programming Reference
--------------------------------

These are just a few highlights. A more detailed reference can be found [here](http://www.stuffaboutcode.com/p/minecraft-api-reference.html)

###World

####`world.getBlock(x, y, z)`
Look up the type of block at the specified coordinates.

####`world.setBlock(x, y, z, block_type)`
Set the block at the specified coordinates to the type block_type.

####`world.setBlocks(x1, y1, z1, x2, y2, z2, block_type)`
Create a set of blocks starting at one coordinate point extending to another point with blocks of the type block_type. This can be used to make cubes or rectangles.

####`world.getHeight(x, z)`
Look up the height (y coordinate) of the tallest brick at the specified x and y coordinates.

####`world.postToChat(“Message”)`
Send a message over chat.

###Player

####`player.getPos()`
Look up the coordinates that the player is currently positioned at.

####`player.setPos(x,y,z)`
Set the player’s position to the specified coordinates.

###Blocks
[mcpi/block.py](https://github.com/sarahhagstrom/python-minecraft/blob/master/mcpi/block.py) is the ultimate authority on which blocks are available for your use. Those code constants are copied and pasted here for convenience. The underlying engine actually provides all of [these](http://minecraft.gamepedia.com/Data_values_(Pocket_Edition)) blocks, not all of which are included in the MCPi block module. You are free to edit mcpi/block.py to add more blocks using the decimal value listed and following the pattern you find in [block.py](https://github.com/sarahhagstrom/python-minecraft/blob/master/mcpi/block.py).

    AIR
    STONE
    GRASS
    DIRT
    COBBLESTONE
    WOOD_PLANKS
    SAPLING
    BEDROCK
    WATER_FLOWING
    WATER
    WATER_STATIONARY
    LAVA_FLOWING
    LAVA
    LAVA_STATIONARY
    SAND
    GRAVEL
    GOLD_ORE
    IRON_ORE
    COAL_ORE
    WOOD
    LEAVES
    GLASS
    LAPIS_LAZULI_ORE
    LAPIS_LAZULI_BLOCK
    SANDSTONE
    BED
    COBWEB
    GRASS_TALL
    WOOL
    FLOWER_YELLOW
    FLOWER_CYAN
    MUSHROOM_BROWN
    MUSHROOM_RED
    GOLD_BLOCK
    IRON_BLOCK
    STONE_SLAB_DOUBLE
    STONE_SLAB
    BRICK_BLOCK
    TNT
    BOOKSHELF
    MOSS_STONE
    OBSIDIAN
    TORCH
    FIRE
    STAIRS_WOOD
    CHEST
    DIAMOND_ORE
    DIAMOND_BLOCK
    CRAFTING_TABLE
    FARMLAND
    FURNACE_INACTIVE
    FURNACE_ACTIVE
    DOOR_WOOD
    LADDER
    RAIL
    STAIRS_COBBLESTONE
    DOOR_IRON
    REDSTONE_ORE
    SNOW
    ICE
    SNOW_BLOCK
    CACTUS
    CLAY
    SUGAR_CANE
    FENCE
    GLOWSTONE_BLOCK
    BEDROCK_INVISIBLE
    STONE_BRICK
    GLASS_PANE
    MELON
    FENCE_GATE
    GLOWING_OBSIDIAN
    NETHER_REACTOR_CORE

Minecraft Controls
------------------

###Keyboard
<table><tbody>
<tr><td>W,A,S,D</td><td>Move (navigate inventory)</td></tr>
<tr><td>SPACE</td><td>Jump, double tap to start/stop flying, hold to fly higher</td></tr>
<tr><td>SHIFT</td><td>Sneak, hold to fly lower</td></tr>
<tr><td>E</td><td>Open inventory</td></tr>
<tr><td>18</td><td>Select inventory slot item to us</td></tr>
<tr><td>ESC</td><td>Show/hide menu</td></tr>
<tr><td>TAB</td><td>Release mouse without showing menu</td></tr>
<tr><td>ENTER</td><td>Confirm menu selection</td></tr>
</tbody></table>

###Mouse
<table><tbody>
<tr><td>Steer</td><td>Look/turn around</td></tr>
<tr><td>Left button</td><td>Remove block (hold)</td></tr>
<tr><td>Right button</td><td>Place block, hit block with sword</td></tr>
<tr><td>Mouse wheel</td><td>Select inventory slot item to use</td></tr>
</tbody></table>

Resources and Credits
----------------------

###Online Coding Classes
[Codecademy](http://www.codecademy.com/)<br>
[LearnPython.org](http://www.learnpython.org/)<br>
[Udacity](https://www.udacity.com/)<br>

###More about the Minecraft API
[MCPi API Basics](http://www.stuffaboutcode.com/2013/01/raspberry-pi-minecraft-api-basics.html)<br>
[MCPi API Tutorial on StuffAboutCode.com](http://www.stuffaboutcode.com/2013/04/minecraft-pi-edition-api-tutorial.html)<br>
[MCPi API Reference](http://www.stuffaboutcode.com/p/minecraft-api-reference.html)<br>
[MCPi API Examples](http://www.stuffaboutcode.com/2013/02/raspberry-pi-minecraft-install.html)<br>
[More MCPi scripts](https://github.com/brooksc/mcpipy)<br>

Note that although the StuffAboutCode links discuss using an Raspberry Pi, you don't need a Raspberry Pi to use the API. We're using CraftBukkit's local server instead. More information about that is [here](/SETUP_README.md)



docs/release-notes.rst
===============
 Release Notes
===============

The following changes have been made to the classroom materials over
time. Look for annotated Git tags that match the headings below.

The `full changelog`_ is available on GitHub.

.. _full changelog: https://github.com/CoderDojoTC/python-minecraft/commits/master

Unreleased Changes
==================

* Merged several documentation fixes from `Markus Khouri`_. Thanks
  Markus!

.. _Markus Khouri: https://github.com/markuskhouri

* Began migrating documentation into the Jupyter/IPython
  notebook. This gives the students a more integrated path through the
  course, with less jumping between the static docs and the live
  environment.

* Added some tools to help with course content development:

  * Added a :file:`requirements.txt` file for use in setting up a
    Python environment with the necessary tools.

  * Added a `Fabric`_ script (:file:`fabfile.py`) to help automate
    tasks performed while developing and delivering the course.

.. _Fabric: http://www.fabfile.org/


2015-08-01
==========

* Fixed the lab server controler to work with the latest version of
  gspread, which requires OAuth2.

* Changed the source location where we pull the Canary JAR file. The
  CanaryMod project is going through some changes, and have
  reorganized where files are available on their web server.

* Added a dirty hack to work around a problem that Canary seems to be
  having in determining the UUID of the players listed in ops.txt on
  startup. I've had to put this reactive code in the run-canary.sh
  script to do the job after the player joins the server. A better
  long-term fix is to find the root cause in Canary itself.


2015-05-09
==========

* Upgraded to CanaryMod 1.2.0 (compatible with Minecraft client
  version 1.8), version 1.3 of the RaspberryJuice plugin, and the
  latest python libraries that support them. This brings some new
  programming capabilities, such as being able to have Python code
  receive messages from the game chat.

* Added Martin O'Hanlon's *Minecraft Turtles* so that students can use
  the library and run the examples. This gives students a logo-like
  model for writing programs.

* Upgraded to IPython 3.1.0. This improves the development user
  interface. It also changes the "branding" of the development UI from
  "IPython" to "Jupyter".

* Removed many calls to the obsolete ``server`` module from example
  scripts in the :file:`examples` folder (thanks to Joshua Fasching).

* Fixed a problem where it seemed to rain in Minecraft from time to
  time.


2015-02-21
==========

* Simplify the generated world in the student lab instance to make it
  better for programming projects. Now, the students are placed in a
  "FLAT" world (no terrain, no biomes, no structures, no mobs, etc.),
  and the weather is clear.

* Reorganized the exercises and examples into subfolders, and
  converted all the previous :file:`*.py` files into :file:`*.ipynb`
  files so students can load them directly in IPython.

* Improved our classroom exercises:

  * **Exercise 1:** Reformatted Hello World to be a single code block
    so that it's easier to run, added some hints, and added some extra
    challenges.

  * **Exercise 2:** Took out the TNT Pyramid (which was too big and
    distracting), pointed students to the online help and guided tour,
    wrote more about added the REPL environment.

  * **Exercise 4:** Added a delay so students have a chance to observe
    what is happening as the script runs, added some extra challenges.

  * **Exercise 5:** Added a field of wool so students have something
    that will react to sword strikes.

* Updated some example scripts to work better in our new student
  environment.

* Upgraded to CanaryRaspberryJuice version 1.2. This added methods
  such as getPlayerId(playerName), getDirection(), getRotation(), and
  getPitch(). This also includes version 1.1, which added various
  entity methods.

* Upgraded to IPython version 2.4 in the student lab server
  instance. This is a significant upgrade from the prior version,
  which was IPython 1.3. The main student-visible differences are that
  IPython can now navigate between directories, and the Notebook
  web UI now has separate Edit and Command modes.

* The docs for setting up a local, Vagrant-based lab environment have
  been updated to work much better. It now uses exactly the same
  environment and setup that students get when attending a CoderDojoTC
  event. The solo-server docs and files have been removed.

* The Lab Server Controller now accepts the name of a docker image
  (and version tags) in its configuration file. This makes it possible
  to use different lab images in different environments, which is
  particularly useful in testing a new image before committing to
  using it for a specific CoderDojo event.

* Lots of documentation updates, including new docs for Mentors.

* For people working on updating exercise or example code through the
  IPython notebook interface, launching a student lab instance with
  Vagrant will now copy the local working directory into the folder
  where IPython can edit the files. If there is no Vagrant-provided
  folder, the code will be retrieved from the Git repository specified
  in the :file:`private_config.yaml` file.

  By itself, this copy from Vagrant into the IPython notebook folder
  is a one-way-street. Edits made there will not be visible on the
  Vagrant host (where they can be committed to source code
  control). However, there is also a script installed in the image
  called :command:`sync-notebooks.sh`. It can be run as follows from
  the Vagrant host, and it will invoke Unison to sync up the changes::

    docker exec -it python-minecraft_default_1424407654 sync-notebooks.sh


2015-02-07
==========

* The table of contents and navigation between chapters has been
  improved. Some of the section index pages have been improved.

* Improved the layout of the Lab Instance Connection card, and its

mcpi/block.py
class Block:
    """Minecraft PI block description. Can be sent to Minecraft.setBlock/s"""
    def __init__(self, id, data=0):
        self.id = id
        self.data = data

    def __cmp__(self, rhs):
        return hash(self) - hash(rhs)

    def __hash__(self):
        return (self.id << 8) + self.data

    def withData(self, data):
        return Block(self.id, data)

    def __iter__(self):
        """Allows a Block to be sent whenever id [and data] is needed"""
        return iter((self.id, self.data))
        
    def __repr__(self):
        return "Block(%d, %d)"%(self.id, self.data)

AIR                 = Block(0)
STONE               = Block(1)
GRASS               = Block(2)
DIRT                = Block(3)
COBBLESTONE         = Block(4)
WOOD_PLANKS         = Block(5)
SAPLING             = Block(6)
BEDROCK             = Block(7)
WATER_FLOWING       = Block(8)
WATER               = WATER_FLOWING
WATER_STATIONARY    = Block(9)
LAVA_FLOWING        = Block(10)
LAVA                = LAVA_FLOWING
LAVA_STATIONARY     = Block(11)
SAND                = Block(12)
GRAVEL              = Block(13)
GOLD_ORE            = Block(14)
IRON_ORE            = Block(15)
COAL_ORE            = Block(16)
WOOD                = Block(17)
LEAVES              = Block(18)
GLASS               = Block(20)
LAPIS_LAZULI_ORE    = Block(21)
LAPIS_LAZULI_BLOCK  = Block(22)
SANDSTONE           = Block(24)
BED                 = Block(26)
COBWEB              = Block(30)
GRASS_TALL          = Block(31)
WOOL                = Block(35)
FLOWER_YELLOW       = Block(37)
FLOWER_CYAN         = Block(38)
MUSHROOM_BROWN      = Block(39)
MUSHROOM_RED        = Block(40)
GOLD_BLOCK          = Block(41)
IRON_BLOCK          = Block(42)
STONE_SLAB_DOUBLE   = Block(43)
STONE_SLAB          = Block(44)
BRICK_BLOCK         = Block(45)
TNT                 = Block(46)
BOOKSHELF           = Block(47)
MOSS_STONE          = Block(48)
OBSIDIAN            = Block(49)
TORCH               = Block(50)
FIRE                = Block(51)
STAIRS_WOOD         = Block(53)
CHEST               = Block(54)
DIAMOND_ORE         = Block(56)
DIAMOND_BLOCK       = Block(57)
CRAFTING_TABLE      = Block(58)
FARMLAND            = Block(60)
FURNACE_INACTIVE    = Block(61)
FURNACE_ACTIVE      = Block(62)
DOOR_WOOD           = Block(64)
LADDER              = Block(65)
RAIL                = Block(66)
STAIRS_COBBLESTONE  = Block(67)
DOOR_IRON           = Block(71)
REDSTONE_ORE        = Block(73)
SNOW                = Block(78)
ICE                 = Block(79)
SNOW_BLOCK          = Block(80)
CACTUS              = Block(81)
CLAY                = Block(82)
SUGAR_CANE          = Block(83)
FENCE               = Block(85)
GLOWSTONE_BLOCK     = Block(89)
BEDROCK_INVISIBLE   = Block(95)
STONE_BRICK         = Block(98)
GLASS_PANE          = Block(102)
MELON               = Block(103)
FENCE_GATE          = Block(107)
GLOWING_OBSIDIAN    = Block(246)
NETHER_REACTOR_CORE = Block(247)
